### PR TITLE
Added RetryPolicy

### PR DESCRIPTION
**Pull request by stankovski**
* Added RetryPolicy implementation
* Refactored NamespaceManager to be closed to NotificationHubClient
* Added tests
* Added ability to set HttpClient in ctors to support IHttpClientFactory
* Removed dead code

### CODE REVIEW

File: src/Microsoft.Azure.NotificationHubs.DotNetCore.Tests/Microsoft.Azure.NotificationHubs.DotNetCore.Tests.csproj [hunk 1]
```diff
@@ -11,6 +11,7 @@
     <PackageReference Include="Microsoft.Extensions.Configuration" Version="3.1.1" />
     <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="3.1.1" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
+    <PackageReference Include="RichardSzalay.MockHttp" Version="6.0.0" />
```
[the remark this answers is not in the document]
Comment (stankovski) [reply src/Microsoft.Azure.NotificationHubs.DotNetCore.Tests/Microsoft.Azure.NotificationHubs.DotNetCore.Tests.csproj:14]: Yes, it's pretty popular for mocking httpclient traffic. We use it in other NH projects.

File: src/Microsoft.Azure.NotificationHubs/NotificationHubRetryOptions.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        private TimeSpan _delay = TimeSpan.FromSeconds(1);
+
+        /// <summary>The maximum delay to allow between retry attempts.</summary>
+        private TimeSpan _maxDelay = TimeSpan.FromMinutes(1);
+
+        /// <summary>
+        ///   The approach to use for calculating retry delays.
+        /// </summary>
+        ///
+        public NotificationHubRetryMode Mode { get; set; } = NotificationHubRetryMode.Exponential;
```
[the remark this answers is not in the document]
Comment (stankovski) [reply src/Microsoft.Azure.NotificationHubs/NotificationHubRetryOptions.cs:31]: This is a pattern used by other Azure libs. It makes configuring the library easier where you need to just set primitives as opposed to initializing complex objects.

File: src/Microsoft.Azure.NotificationHubs.DotNetCore.Tests/NamespaceManagerRetryPolicyTests.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved. 
+// Licensed under the MIT License. See License.txt in the project root for 
+// license information.
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.NotificationHubs.DotNetCore.Tests
+{
+    using System;
```
Comment (mattpodwysocki) [inline src/Microsoft.Azure.NotificationHubs.DotNetCore.Tests/NamespaceManagerRetryPolicyTests.cs:9]: Let's standardize with usings outside of namespaces as per our coding standards

File: src/Microsoft.Azure.NotificationHubs/NotificationHubRetryOptions.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        /// <summary>
+        ///   A custom retry policy to be used in place of the individual option values.
+        /// </summary>
+        ///
+        /// <remarks>
+        ///   When populated, this custom policy will take precedence over the individual retry
+        ///   options provided.
+        /// </remarks>
+        ///
+        public NotificationHubRetryPolicy CustomRetryPolicy { get; set; }
```
[the remark this answers is not in the document]
Comment (stankovski) [reply src/Microsoft.Azure.NotificationHubs/NotificationHubRetryOptions.cs:101]: See above

File: src/Microsoft.Azure.NotificationHubs/BasicRetryPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+        /// <summary>
+        ///   The set of options responsible for configuring the retry
+        ///   behavior.
+        /// </summary>
+        ///
+        public NotificationHubRetryOptions Options { get; }
+
+        /// <summary>
+        ///   The factor to apply to the base delay for use as a base jitter value.
```
[the remark this answers is not in the document]
Comment (stankovski) [reply src/Microsoft.Azure.NotificationHubs/BasicRetryPolicy.cs:37]: Any suggestions how to make this more descriptive? I've copied this from Azure SB lib

[the remark this answers is not in the document]
Comment (stankovski) [reply src/Microsoft.Azure.NotificationHubs/BasicRetryPolicy.cs:37]: Updated based on suggestion

**Pull request merged**